### ISSUE TITLE
ECDSA validation not enforced

### ISSUE THREAD

**Issue by dilyanpalauzov** [matteocorti/check_ssl_cert#180]
`openssl s_client -connect imap.gmail.com:993 -sigalgs 'ECDSA+SHA1:ECDSA+SHA224:ECDSA+SHA384:ECDSA+SHA256:ECDSA+SHA512'` prints
```
depth=2 OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign
verify return:1
depth=1 C = US, O = Google Trust Services, CN = GTS CA 1O1
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = imap.gmail.com
verify return:1
140072184799744:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1543:SSL alert number 40
CONNECTED(00000003)
…
```
so it does not work and returns 1.  `openssl s_client -connect imap.gmail.com:993 -sigalgs 'RSA-PSS+SHA512:RSA-PSS+SHA384:RSA-PSS+SHA256:RSA+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA224:RSA+SHA1'` does work and returns 0.

But
```
./check_ssl_cert  -H imap.gmail.com -p 993 --ecdsa
./check_ssl_cert  -H imap.gmail.com -p 993 --rsa
```
both return 0 and both state `SSL_CERT OK - x509 certificate 'imap.gmail.com' from 'GTS CA 1O1' valid until Jan  2 20:57:36 2020 GMT (expires in 68 days)|days=68;;;;`.

`./check_ssl_cert  -H imap.gmail.com -p 993 --ecdsa` must fail but does not.

**Comment by dilyanpalauzov** [comment]
This enforces failing, albeit with unclear error message:
```diff
diff --git a/check_ssl_cert b/check_ssl_cert
--- a/check_ssl_cert
+++ b/check_ssl_cert
@@ -475,8 +475,11 @@ exec_with_timeout() {
 
         eval "${TIMEOUT_BIN} ${time} ${command}" > /dev/null 2>&1
 
-        if [ $? -eq 137 ] ; then
+        RET=$?
+        if [ $RET -eq 137 ] ; then
             prepend_critical_message "Timeout after ${time} seconds"
+        elif [ $RET -eq 1 ] ; then
+           prepend_critical_message FAILURE
         fi
 
     elif [ -n "${EXPECT}" ] ; then
```
On my system with coreutils 8.31, timeout --help says:
```
If the command times out, and --preserve-status is not set, then exit with
status 124.
```
so the code shall check for 124 to identify timeout, but it looks for 137.

**Comment by dilyanpalauzov** [comment]
A better fix:
```diff
diff --git a/check_ssl_cert b/check_ssl_cert
--- a/check_ssl_cert
+++ b/check_ssl_cert
@@ -475,9 +475,11 @@ exec_with_timeout() {
 
         eval "${TIMEOUT_BIN} ${time} ${command}" > /dev/null 2>&1
 
-        if [ $? -eq 137 ] ; then
+        RET=$?
+        if [ $RET -eq 137 -o $RET -eq 124 ] ; then
             prepend_critical_message "Timeout after ${time} seconds"
         fi
+        return $RET
 
     elif [ -n "${EXPECT}" ] ; then
 
```
With it applied, `./check_ssl_cert -H imap.gmail.com -p 993 --ecdsa` prints `SSL_CERT CRITICAL imap.gmail.com: SSL error: 140052905800192:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1543:SSL alert number 40`

I do not know how to dissect this further.  In particular I do not understand why other hosts not offering RSA do fail in another way.  In any case, this fix does force ./check_ssl_cert to fail, when ECDSA is requested, but not offered.

**Comment by matteocorti** [comment]
On my version I have

```
If the command times out, and --preserve-status is not set, then exit with
status 124.  Otherwise, exit with the status of COMMAND.  If no signal
is specified, send the TERM signal upon timeout.  The TERM signal kills
any process that does not block or catch that signal.  It may be necessary
to use the KILL (9) signal, since this signal cannot be caught, in which
case the exit status is 128+9 rather than 124.
```

I will check

**Comment by dilyanpalauzov** [comment]
https://www.gnu.org/software/coreutils/manual/coreutils.html#timeout-invocation has one more explanation for the return valuej

**Comment by matteocorti** [comment]
I will fix the return codes but the problem should be analysed also on systems not using timeout

**Comment by matteocorti** [comment]
```
printf 'HEAD / HTTP/1.1\nHost: imap.gmail.com\nUser-Agent: check_ssl_cert/1.102.0\nConnection: close\n\n' | /opt/local/bin/openssl s_client    -crlf -ign_eof -connect imap.gmail.com:993 -servername imap.gmail.com -showcerts -verify 6    -sigalgs 'ECDSA+SHA1:ECDSA+SHA224:ECDSA+SHA384:ECDSA+SHA256:ECDSA+SHA512' 
```

generates a return code of `1` but the plugin gives an OK value

**Comment by dilyanpalauzov** [comment]
I am not sure how to interpret the last comment.  That last command prints on STDERR
```
verify depth is 6
depth=2 OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign
verify return:1
depth=1 C = US, O = Google Trust Services, CN = GTS CA 1O1
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = imap.gmail.com
verify return:1
140316498592256:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1543:SSL alert number 40
```

and the top-most certificate, when stored in a file and viewed with `openssl x509 -in THE-FILE -noout -text` prints
```
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
```
so the returned certificate is RSA, not EC.

**Comment by matteocorti** [comment]
The comment means that the problem is not bound to the usage of timeout. On my mac I do not have timeout and the plugins generates an OK (which is not OK).

Your fix, fixes the problem with timeout, but not with expect. I'll take a look

**Comment by dilyanpalauzov** [comment]
I asked at https://github.com/openssl/openssl/issues/10131#issuecomment-546962162 for assistance in the interpretation.

**Comment by matteocorti** [comment]
What I meant is that at the moment executing the command with expect completely ignores the command return code

**Comment by dilyanpalauzov** [comment]
How does expect determine failure and success?

**Comment by matteocorti** [comment]
At the moment it is ignored.

I'm taking a look at
https://stackoverflow.com/questions/23614039/how-to-get-the-exit-code-of-spawned-process-in-expect-shell-script

**Comment by matteocorti** [comment]
I now fixed (4f71216cf5cb96ab01832bcd8823260e658abe22) both the `timeout` and `expect` implementations so that the return value is no more ignores. Now I'll take a look at the actual issue.

**Comment by matteocorti** [comment]
Sorry it's fixed in (a6849a9cf4ca13d437ff62ce833e6c0feeab679f)

Now

```
$ ./check_ssl_cert -H imap.gmail.com -p 993 --ecdsa 
SSL_CERT CRITICAL imap.gmail.com: SSL error: 4589698496:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1544:SSL alert number 40
```

**Comment by matteocorti** [comment]
The issue is now fixed. The error message is not so nice but the parsing of `s_client` messages should be handled in a general way and separately.
